# Practice 3

In this exercise, you will practice wrangling data through different formats, with missing data, and working with text.

You can either print answers directly from your code or write them in the markdown cells below your code. Either way, make sure that your answers are visible and can be easily read in the final notebook you turn in.

## Part 1: Text wrangling and regular expressions 

### Question 1 (3 points)
Below we import the `text.txt` file as a string called `text` and preview the first 1,000 characters. Note that the text has no capitalization at the beginning of sentences.

In [12]:
# Run but do not modify this code
with open("text.txt") as f:
    text = f.read()

print(text[:1000])

it was the best of times, it was the worst of times, it was the age of wisdom, it was the age of foolishness, it was the epoch of belief, it was the epoch of incredulity, it was the season of light, it was the season of darkness, it was the spring of hope, it was the winter of despair, we had everything before us, we had nothing before us, we were all going direct to heaven, we were all going direct the other way- in short, the period was so far like the present period, that some of its noisiest authorities insisted on its being received, for good or for evil, in the superlative degree of comparison only. there were a king with a large jaw and a queen with a plain face, on the throne of england; there were a king with a large jaw and a queen with a fair face, on the throne of france. in both countries it was clearer than crystal to the lords of the state preserves of loaves and fishes, that things in general were settled for ever. it was the year of our lord one thousand seven hundred 

Capitalize the first letter of each sentence (e.g., "it was the best..." should become "It was the best..."). To do so, split the sentences (you may assume sentences are ended with the a period `.` followed by a blank space), capitalize the first letter of each sentence, and then join the results.

When you finish, print the first 1,000 characters. You do not need to write anything else in your answer.

In [30]:
newtext=text.split(".")
newtext=[x.lstrip().capitalize() for x in newtext]
newtext=". ".join(arr)
print(newtext[:1000])





It was the best of times, it was the worst of times, it was the age of wisdom, it was the age of foolishness, it was the epoch of belief, it was the epoch of incredulity, it was the season of light, it was the season of darkness, it was the spring of hope, it was the winter of despair, we had everything before us, we had nothing before us, we were all going direct to heaven, we were all going direct the other way- in short, the period was so far like the present period, that some of its noisiest authorities insisted on its being received, for good or for evil, in the superlative degree of comparison only. There were a king with a large jaw and a queen with a plain face, on the throne of england; there were a king with a large jaw and a queen with a fair face, on the throne of france. In both countries it was clearer than crystal to the lords of the state preserves of loaves and fishes, that things in general were settled for ever. It was the year of our lord one thousand seven hundred 

### Question 2 (4 points)
Below we import the `corrupted_text.txt` file, a messed up version of `text.txt`, and print the first 1,000 characters.

In [32]:
# Run but do not modify this code
with open("corrupted_text.txt") as f:
    ctext = f.read()

print(ctext[:1000])

it was the best of times$ it was the worst of times$####                           it was the age of wisdom$ it was the age of foolishness$ it was the epoch of belief$ it was the epoch of incredulity$ it was the season of light$ it was the season of darkness$ it was the spring of hope$ it was the winter of despair$####                  we had everything before us$####          we had nothing before us$ we were all going direct to heaven$ we were all going direct the other way- in short$ the period was so far like the present period$ that some of its noisiest authorities insisted on its being received$ for good or for evil$ in the superlative degree of comparison only$\n\n there were a king with a large jaw and a queen with a plain face$ on the throne of england; there were a king with a large jaw and a queen with a fair face$ on the throne of france$\n\n in both countries it was clearer than crystal to the lords of the state preserves of loaves and fishes$ that things in general were s

Fix the following problems to recover the original text that appeared in `text.txt`.

1. Every comma `,` has been replaced by `$`. Sometimes, but not always, this is followed by `####`. When it is followed by `####`, there are different numbers of blank spaces after the `####`.
2. Every period `.` has been replaced by `$\n\n`.

When you are finished, print the first 1,000 characters. You should see the same text as in `text.txt` above (except without the capitalization you worked on in the previous problem).  You do not need to write anything else in your answer.

Note: It may be helpful to recall that `\` is an escape character for Python strings, so for example, while calling `print("\n")` will simply print a blank newline character, calling `print("\\n")` will "escape" the usual meaning of the second backslash and print `\n`.

In [81]:
newertext=ctext.split("$\\n\\n")
newertext=[x.strip("\\n, ,,#,$") for x in newertext]
newertext=". ".join(newertext)
newertext=newertext.split("$")
newertext=[j.strip(" ,,#") for j in newertext]
newertext=", ".join(newertext)
print(newertext[:1000])



it was the best of times, it was the worst of times, it was the age of wisdom, it was the age of foolishness, it was the epoch of belief, it was the epoch of incredulity, it was the season of light, it was the season of darkness, it was the spring of hope, it was the winter of despair, we had everything before us, we had nothing before us, we were all going direct to heaven, we were all going direct the other way- in short, the period was so far like the present period, that some of its noisiest authorities insisted on its being received, for good or for evil, in the superlative degree of comparison only. there were a king with a large jaw and a queen with a plain face, on the throne of england; there were a king with a large jaw and a queen with a fair face, on the throne of france. in both countries it was clearer than crystal to the lords of the state preserves of loaves and fishes, that things in general were settled for ever. it was the year of our lord one thousand seven hundred 

In [2]:
# Run but do not modify this code
import re # Regular expression library
with open("text.txt") as f:
    ftext = f.read()

Write regular expressions and use the Python `re` module [documentation](https://docs.python.org/3/library/re.html) to answer the following questions about the file. You may assume that all characters are lowercase. You can print your answers directly from code (as long as they are clearly labeled) or you can write them in the markdown cell below.

1. How many times do two vowels (a, e, i, o, and u) appear consecutively in `text`? When three vowels occur consecutively (such as in "queen") you only need to count once, (for example, in "queen" you do not need to worry about the overlap between "ue" and "ee", just count once). 
2. Print all of the hyphenated words in `text` (that is, the words with a hyphen character `-` in the middle and no spaces between the hyphen and the characters on either side). 
3. Print the first word that appears after each period character `.` in the text. Do not include the period itself or the space that follows it.
4. Print the starting index of every word `she` or `he`. Do not include words that enclose these such as `the` or `fishes` (Hint: It appears only 3 times).

In [4]:
# Put your code to answer the question here
query1 = re.compile(r"[aeiou][aeiou]")
print(len(re.findall(query1, ftext)))
query2=re.compile(r'[A-Za-z]+[-]+[A-Za-z]+')
print(re.findall(query2, ftext))
query3=re.compile(r'\. ([A-Za-z]+)')
print(re.findall(query3, ftext))
query5=re.compile(r'[" "]+([she]+)[" "]+')
arry=re.finditer(query5, ftext)
for i in arry:
    print(i)
print()

184
['seventy-five', 'five-and', 'cock-lane', 'cock-lane', 'fellow-tradesman', 'drawing-rooms', 'to-day', 'to-morrow', 'seventy-five', 'seventy-five', 'creatures-the', 'rest-along']
['there', 'in', 'it', 'spiritual', 'mrs', 'southcott', 'even', 'mere', 'france', 'under', 'it', 'it', 'but', 'in', 'daring', 'giles', 'in', 'all', 'environed', 'thus']
<re.Match object; span=(2086, 2091), match=' she '>
<re.Match object; span=(2268, 2272), match=' he '>
<re.Match object; span=(3759, 3763), match=' he '>



### Answer 3
1. 184
2. ['seventy-five',
 'five-and',
 'cock-lane',
 'cock-lane',
 'fellow-tradesman',
 'drawing-rooms',
 'to-day',
 'to-morrow',
 'seventy-five',
 'seventy-five',
 'creatures-the',
 'rest-along']
3. ['there', 'in', 'it', 'spiritual', 'mrs', 'southcott', 'even', 'mere', 'france', 'under', 'it', 'it', 'but', 'in', 'daring', 'giles', 'in', 'all', 'environed', 'thus']
4. [2087, 2269, 3760]

## Part 2: Cleaning up system logs CSV
In this part we work with a piece of messy tabular data in the form of a poorly formatted csv file containing data about programs running on computer systems. 

### Question 4 (6 points)
Below, we import the dataset using the Pandas `read_csv` function that creates a dataframe. Run the code; it will preview the first five rows. There are several formatting issues with the default import.

In [7]:
# Run but do not modify this code
import pandas as pd
sys_df = pd.read_csv("monitor.csv")
sys_df.head()

0  System Time: 23 second   System Mem: 3 Gb  414 Mb   
0  1  System Time: 40 second   System Mem: 3 Gb  382 Mb   
1  2  System Time: 43 second   System Mem: 2 Gb  271 Mb   
2  3  System Time: 31 second   System Mem: 3 Gb  493 Mb   
3  4  System Time: 37 second   System Mem: 3 Gb  411 Mb   
4  5  System Time: 36 second   System Mem: 2 Gb  266 Mb

Address the following, either by changing the `read_csv` statement or by modifying the imported dataframe:

1. The rows are already numbered with an integer index, but that is being duplicated so that we see a primary index for the dataframe and separately a column containing integer indices. Fix this so that there is just a single integer index as the primary index of the dataframe (bolded).
2. There are commas at the end of every row causing the dataframe to add a blank extra column at the right. Get rid of this column.
3. The file did not have column labels, so currently the first row is being incorrectly interpreted as the bolded column labels. Fix the code so that the column labels (from left to right) are `System Time second`, `System Memory GB`, and `System Memory MB`.

When you are finished, `sys_df` should have the above issues correct. Again run `sys_df.head()` to show the first few rows of your dataframe. You do not need to write anything else in your answer.

In [8]:
# Put your code to answer the question here
# You are welcome to use a new sys_df = pd.read_csv(...) statement

### Question 5 (6 points)
The `sys_df` dataframe from question 3 should now be a little easier to read and use, but there are some strange things we would like to fix about the formatting of the data.

1. The three features or columns all contain numeric data (seconds, GB, and MB) but are currently formatted as Strings with redundant information also contained in the column names. Fix the dataframe so that each value is just an integer. Note that the rows at index `400` and on have System Time recorded in minutes instead of seconds, be sure to convert these to seconds by mulitplying by 60.
2. Currently the System Memory is split accross two columns, one for the GB and one for the MB. For example, the total memory of the first program is 3 GB and 414 MB. Instead, represent the full system memory in the `System Memory MB` column, and get rid of the `System Memory GB` column. To do so, you need to convert the values in the GB column to MB (1 GB is 1000 MB) and add that to the MB column, then use the [`drop` method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) to remove the GB column.
3. Now that you have dealt with the above, it should be easy to provide some basic statistics about the data. Report the mean time in seconds and mean memory in MB of the programs.

When you are finished, again run `sys_df.head()` to show the first few rows of your dataframe, and print or report your answer to the third task.

In [9]:
# Put your code to answer the question here

### Answer 5
Write your answer here if not printed above 

## Part 3: Wrangling FDA JSON Dataset 
In this part we work with a messy JSON dataset containing information about several drugs labels.

### Question 6 (6 points)
Below we import the `FDADrugLabel.json` file into the `labels` variable.

The resulting Python object is somewhat messy; we encourage you to explore the data before answering the questions. For example, rather than just printing out `labels` check it's type, print only one element, and remember there are different [dictionary functions](https://docs.python.org/3/library/stdtypes.html#typesmapping) you could use. Feel free to add your own cells as you explore.

In [10]:
# Run but do not modify this code
import json
with open("FDADrugLabel.json") as f:
    labels = json.load(f)

Answer the following questions.

1. How many drugs are represented in the file?
2. What is the largest number of key/value (or name/value) pairs for any drug in the file?
3. The `effective_time` is recorded for each drug as a date in the format `yyyymmdd`, for example, June 1, 1992 would be recorded as `19920601`. What is the earliest recorded date?

In [11]:
# Put your code to answer the question here

### Answer 6
Write your answer here if not printed above

### Question 7 (6 points)

Below we use the Pandas `read_json` function to import the `FDADrugLabel.json` file into a Pandas dataframe `label_df`.

In [12]:
# Run but do not modify this code
import pandas as pd
label_df = pd.read_json("FDADrugLabel.json")
label_df.head(20)

package_label_principal_display_panel  \
0                                            [label]   
1                           [BrightSpark Oral Spray]   
2  [PRINCIPAL DISPLAY PANEL - 106 g Packet Carton...   
3  [PACKAGE LABEL-PRINCIPAL DISPLAY PANEL NDC 631...   
4  [PACKAGE LABEL DISPLAY: ENERGIQUE SINCE 1987 H...   

                                   active_ingredient  \
0                                   titanium dioxide   
1  HPUS Active Ingredients Each dose contains equ...   
2            Active ingredient Colloidal oatmeal 43%   
3                                                NaN   
4  ACTIVE INGREDIENTS: Angelica Archangelica 3X, ...   

                                            warnings  \
0  [1. Do not use in the following cases(Eczema a...   
1  [Warnings If symptoms persist or worsen, consu...   
2  [Warnings For external use only. When using th...   
3                                                NaN   
4  [WARNINGS: If pregnant or breastfeeding, ask a...   

                                 inactive_ingredient  \
0                        butylene glycol, water, etc   
1    Citric acid, potassium sorbate, purified water.   
2  Inactive ingredients mineral oil, calcium sili...   
3                                                NaN   
4  INACTIVE INGREDIENTS: Demineralized water, 20%...   

                                   drug_interactions  effective_time  \
0                                                           20160321   
1                                                NaN        20160323   
2                                                NaN        20160315   
3  [7 DRUG INTERACTIONS In vitro data indicate th...        20160301   
4                                                NaN        20160322   

                                             openfda  \
0  {'manufacturer_name': 'Nature and Health Beaut...   
1  {'manufacturer_name': 'Silver Star Brands', 'u...   
2  {'manufacturer_name': 'Johnson & Johnson Consu...   
3  {'manufacturer_name': 'Proficient Rx LP', 'uni...   
4  {'manufacturer_name': 'Energique, Inc.', 'unii...   

                       keep_out_of_reach_of_children  \
0                [keep out of reach of the children]   
1  [Keep this and all medication out of reach of ...   
2  [Keep out of reach of children. If swallowed, ...   
3                                                NaN   
4  [KEEP OUT OF REACH OF CHILDREN: Keep out of re...   

                           spl_product_data_elements  \
0  [Nature and Health Beauty (Moisturizing Cream,...   
1  [BrightSpark Arg nit, Arsenicum iod, Hyoscyamu...   
2  [Aveeno Baby Eczema Therapy Soothing Bath Trea...   
3  [levocetirizine dihydrochloride levocetirizine...   
4  [Ovapar Angelica Archangelica, Caulophyllum Th...   

                                 set_id  ... ask_doctor  do_not_use  \
0  2ea9b21c-18e6-0a8a-e054-00144ff8d46c  ...        NaN         NaN   
1  2eb93d7c-330f-4b93-e054-00144ff88e88  ...        NaN         NaN   
2  7f689b72-ccfb-4ae0-9747-1490593998ca  ...        NaN         NaN   
3  b84ccb27-9c52-43b9-9a46-8f52a571045e  ...        NaN         NaN   
4  13ecc450-0ae3-46d3-8cc7-915c74c3361a  ...        NaN         NaN   

  recent_major_changes boxed_warning indications_and_usage_table precautions  \
0                  NaN           NaN                         NaN         NaN   
1                  NaN           NaN                         NaN         NaN   
2                  NaN           NaN                         NaN         NaN   
3                  NaN           NaN                         NaN         NaN   
4                  NaN           NaN                         NaN         NaN   

  labor_and_delivery general_precautions dosage_and_administration_table  \
0                NaN                 NaN                             NaN   
1                NaN                 NaN                             NaN   
2                NaN                 NaN                             NaN   
3          

Use the dataframe to answer the following questions. 

1. What happens if you call `label_df.dropna()` (with no optional parameters) to remove the null values? Why? 
2. Which of `label_df`'s columns contains data for every row? You might answer this using the `label_df.dropna()` method to keep all of the rows but remove any of the columns that have any missing data, but do not modify the `label_df` dataframe itself. 
3. Filter the dataframe to the rows that are missing data for `active_ingredient`, then show those rows (if `my_view` is a dataframe view you have computed by filtering the data, you can simply put `my_view` as the last line of a cell and run it to get a tabular rendering of the rows in Jupyter). You may find the Pandas `isnull` method useful, you can see it's [documentation here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.isnull.html).

### Answer 7
Write your answer here if not printed above

For item 1, calling `label_df.dropna()` returns a dataframe with no rows, i.e., with all of the data removed. This happens because the default is for `dropna()` to drop every *row* containing *any* missing data, and each row in the `label_df` dataframe has at least one piece of missing data.